### ISSUE TITLE
Custom Template Controller

### ISSUE THREAD

**Issue by dmgawel** [soberwp/controller#37]
Hi!

**TL;DR**
By the way WordPress saves custom page template path and how Sage expands it into possible template files one cannot create Controller based on page template file name.

**Current behavior**
Let's say we have a custom page template located in `views/template-custom.blade.php`. The following is available controller hierarchy:

```
Hierarchy Debugger:
/controllers/app.php
/controllers/index.php
/controllers/resources/views/singular.php
/controllers/singular.php
/controllers/views/singular.php
/controllers/resources/views/page.php
/controllers/page.php
/controllers/views/page.php
/controllers/resources/views/page-4.php
/controllers/page-4.php
/controllers/views/page-4.php
/controllers/resources/views/page-signup.php
/controllers/page-signup.php
/controllers/views/page-signup.php
```

**Expected behavior**

Controller hierarchy should include `/controllers/template-custom.php` and `/controllers/views/template-custom.php`.

**Details**
Let's get back to our custom template located in`views/template-custom.blade.php`. WordPress saves it's exact path and serves as template name. Then Sage's `template_hierarchy` [filter expands it](https://github.com/roots/sage/blob/master/app/helpers.php#L82) into following locations :

```
resources/views/views/template-custom.blade.php
resources/views/views/template-custom.blade.blade.php
views/template-custom.blade.php
views/template-custom.blade.blade.php
views/views/template-custom.blade.php
views/views/template-custom.blade.blade.php
```

I know, it looks broken but it does it's job (the list includes original template path). I'm not sure if this should be 'fixed' by Sage or bypassed by controller.

**Comment by dmgawel** [comment]
I've also opened an issue in Sage: https://github.com/roots/sage/issues/1936

**Comment by alwaysblank** [comment]
Maybe I'm misunderstanding the issue, but I just build a project that made extensive use of controllers for custom page templates. My file organization looked like this:

**Template:**
`resources/views/template-science.blade.php`

**Controller:**
`resources/controllers/views/TemplateScience.php`

That worked as intended, with no other modifications on my part. Is there something I'm missing?

**Comment by dmgawel** [comment]
@alwaysblank that's weird. Could please you add `@debug('hierarchy')` to your template and paste the output?

**Comment by alwaysblank** [comment]
```php
Hierarchy Debugger:
/controllers/app.php
/controllers/singular.php
/controllers/page.php
/controllers/page-14.php
/controllers/page-science.php
/controllers/views/template-science.php
```

[This line](https://github.com/soberwp/controller/blob/master/src/Loader.php#L84):
```php
$template = strtolower(preg_replace('/(?<!^)[A-Z]/', '-$0', $template));
```
in `Loader.php` converts from camelCase to the dash-case WordPress expects so that it can match correctly, which (I think) is why it appears as dash-case in the debugger and not camelCase.

**Comment by dmgawel** [comment]
I've created the files as described by you but it's not working. Here is my hierarchy debug:

```
Hierarchy Debugger:
/controllers/app.php
/controllers/index.php
/controllers/resources/views/singular.php
/controllers/singular.php
/controllers/views/singular.php
/controllers/resources/views/page.php
/controllers/page.php
/controllers/views/page.php
/controllers/resources/views/page-4.php
/controllers/page-4.php
/controllers/views/page-4.php
/controllers/resources/views/page-science.php
/controllers/page-science.php
/controllers/views/page-science.php
```

Looks that there are some differences between our codebases. I'm using Sage dev-master (9.0.0-beta4). What are your versions of Sage and controller?

**Comment by derdeedas** [comment]
Same issue here running sage 9.0.0-beta.4

**Comment by alwaysblank** [comment]
I'm using a slightly modified version of Sage that's pretty close to 9.0.0-beta3. None of my modifications should affect template name resolution, though. Controller is at 9.0.0-beta.3.

**Comment by derdeedas** [comment]
Downgraded Sage & Controller to 9.0.0-beta3, still the same issue 😞

**Comment by darrenjacoby** [comment]
Hi all, 

I will see if I can reproduce and resolve today. It should be working as per @alwaysblank implementation.

**Comment by darrenjacoby** [comment]
So this looks like the change causing the issues is below; 

https://github.com/roots/sage/commit/8691074b29c4c09dae40f0f0c33dc2b51d9292e2#diff-8b748c75763ffb3068f4da1679ecf02f

https://github.com/roots/sage/commit/8691074b29c4c09dae40f0f0c33dc2b51d9292e2#diff-3223ef754af7d30e9a7234fc86ca4eb7

@QWp6t could we look at reverting to return only one instance? It's returning controller files, etc, in the hierarchy. If there is a specific reason for this, I can look at trying to bypass with Controller.

**Comment by alwaysblank** [comment]
I've created a pull request for roots/sage that should resolve this issue: https://github.com/roots/sage/pull/1939 

Please take a look and let me know if it works for you.

**Comment by thomascharbit** [comment]
Had the same issue while it was working with previous Sage version, your pull request fixed it! Thanks!

**Comment by darrenjacoby** [comment]
Thanks @alwaysblank !

**Comment by 9585999** [comment]
@alwaysblank 

Strangely, I inserted the code that you sent to PR and nothing has changed ...

The problem was solved when in the available code I commented the line `"{$path}/{$template}.php"`...

**Comment by 9585999** [comment]
So it works without interference in sage files.

**Comment by Mellis84** [comment]
Cheers @alwaysblank that worked for me also! :) thought I was loosing my mind.

**Comment by darrenjacoby** [comment]
Resolved with https://github.com/roots/sage/commit/499f809c6e3271f1b395c66e7d4b88d54190b73c
